File: python/y2025/m04/lowest_common_ancestor_of_deepest_leaves/lowest_common_ancestor_of_deepest_leaves.py
import time
from typing import List, Optional

'''
    Reescrita em python do problema desenvolvido em java.
'''


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def __init__(self):
        self.max_depth = -1
        self.res = None

    def lcaDeepestLeaves(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        self.res = root
        self.dfs(root, 0)
        return self.res

    def dfs(self, node: Optional[TreeNode], depth: int) -> int:
        if not node:
            self.max_depth = max(self.max_depth, depth)
            return depth

        ld = self.dfs(node.left, depth + 1)
        rd = self.dfs(node.right, depth + 1)

        if ld == rd and ld == self.max_depth:
            self.res = node

        return max(ld, rd)


def main() -> None:
    print("=== Lowest Common Ancestor of Deepest Leaves Tests ===")
    test_cases = [
        ([3, 5, 1, 6, 2, 0, 8, None, None, 7, 4], 2, 1),
        ([1], 1, 2),
        ([0, 1, 3, None, 2], 2, 3),
        ([1, 2, 3, 4, 5, 6, 7], 1, 4),
        ([1, 2, None, 3, None, 4, None], 4, 5)
    ]

    for nodes, expected, test_id in test_cases:
        root = createTree(nodes)
        test_case(root, expected, test_id)


def test_case(root: Optional[TreeNode], expected: int, test_id: int) -> None:
    print(f"\nTest Case {test_id}:\nInput: {treeToString(root)}")

    start_time = time.perf_counter_ns()
    solution = Solution()
    result_node = solution.lcaDeepestLeaves(root)
    result = result_node.val if result_node else None
    elapsed_time = time.perf_counter_ns() - start_time

    print(f"Expected: {expected}")
    print(f"Result:   {result}")
    print(f"Time: {elapsed_time / 1000:.3f} μs")
    print("✅ PASS" if result == expected else "❌ FAIL")


def createTree(nodes: List[Optional[int]]) -> Optional[TreeNode]:
    if not nodes or nodes[0] is None:
        return None

    root = TreeNode(nodes[0])
    queue = [root]
    i = 1

    while queue and i < len(nodes):
        current = queue.pop(0)

        if i < len(nodes) and nodes[i] is not None:
            current.left = TreeNode(nodes[i])
            queue.append(current.left)
        i += 1

        if i < len(nodes) and nodes[i] is not None:
            current.right = TreeNode(nodes[i])
            queue.append(current.right)
        i += 1

    return root


def treeToString(root: Optional[TreeNode]) -> List[str]:
    if not root:
        return []

    result = []
    queue = [root]

    while queue:
        node = queue.pop(0)

        if node:
            result.append(str(node.val))
            queue.append(node.left)
            queue.append(node.right)
        else:
            result.append("null")

    while len(result) > 0 and result[-1] == "null":
        result.pop()

    return result


if __name__ == "__main__":
    main()

File: java/me/study/y2025/m04/lowest_common_ancestor_of_deepest_leaves/LowestCommonAncestorOfDeepestLeaves.java
package me.study.y2025.m04.lowest_common_ancestor_of_deepest_leaves;

import java.util.*;

public class LowestCommonAncestorOfDeepestLeaves {

    /*
        O objetivo é encontrar o LCA que contem as folhas mais profundas da arvore.
        Basta percorrer a arvore, identificar o LCA mais profunto e retorna-lo.

        n = nodes.length
        Time: O(n)
        Space: O(n)
     */

    static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }

    }
    static int maxDepth = -1;
    static TreeNode res = null;

    public static TreeNode lcaDeepestLeaves(TreeNode root) {
        res = root;
        dfs(root, 0);
        return res;
    }


    public static int dfs(TreeNode root, int depth){
        if(root == null){
            maxDepth =  Math.max(depth ,maxDepth);
            return depth;
        }

        int ldmax = dfs(root.left, depth+1);
        int rdmax = dfs(root.right, depth+1);

        System.out.println("processando v:" + root.val + " d:" + depth + " maxd:" + maxDepth + " res: "+ res.val);
        if(ldmax == rdmax && ldmax == maxDepth){
            res = root;
        }
        return Math.max(ldmax, rdmax);
    }



    public static void main(String[] args) {
        System.out.println("=== Lowest Common Ancestor of Deepest Leaves Tests ===");

        // Test Case 1: Example from problem description
        testCase(createTree(new Integer[]{3,5,1,6,2,0,8,null,null,7,4}), 2, 1);

        // Test Case 2: Single node tree
        testCase(createTree(new Integer[]{1}), 1, 2);

        // Test Case 3: Deepest leaf is in right subtree
        testCase(createTree(new Integer[]{0,1,3,null,2}), 2, 3);

        // Test Case 4: Balanced tree with multiple deepest leaves
        testCase(createTree(new Integer[]{1,2,3,4,5,6,7}), 1, 4);

        // Test Case 5: Skewed tree
        testCase(createTree(new Integer[]{1,2,null,3,null,4,null}), 4, 5);
    }

    private static void testCase(TreeNode root, int expected, int testId) {
        System.out.printf("\nTest Case %d:%nInput Tree: %s%n",
                testId, treeToString(root));

        long start = System.nanoTime();
        TreeNode result = lcaDeepestLeaves(root);
        long duration = (System.nanoTime() - start) / 1000;

        System.out.printf("Expected LCA: %d%nResult: %s%n",
                expected, result != null ? result.val : "null");
        System.out.printf("Time: %d μs%n", duration);

        if (result != null && result.val == expected) {
            System.out.println("✅ PASS");
        } else {
            System.out.println("❌ FAIL");
        }
    }

    // Helper method to create tree from level-order traversal array
    private static TreeNode createTree(Integer[] nodes) {
        if (nodes == null || nodes.length == 0 || nodes[0] == null) {
            return null;
        }

        TreeNode root = new TreeNode(nodes[0]);
        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);

        int i = 1;
        while (!queue.isEmpty() && i < nodes.length) {
            TreeNode current = queue.poll();

            if (i < nodes.length && nodes[i] != null) {
                current.left = new TreeNode(nodes[i]);
                queue.add(current.left);
            }
            i++;

            if (i < nodes.length && nodes[i] != null) {
                current.right = new TreeNode(nodes[i]);
                queue.add(current.right);
            }
            i++;
        }

        return root;
    }

    // Helper method to convert tree to string representation
    private static String treeToString(TreeNode root) {
        if (root == null) return "[]";

        List<String> result = new ArrayList<>();
        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);

        while (!queue.isEmpty()) {
            TreeNode node = queue.poll();
            if (node != null) {
                result.add(String.valueOf(node.val));
                queue.add(node.left);
                queue.add(node.right);
            } else {
                result.add("null");
            }
        }

        // Remove trailing nulls
        int i = result.size() - 1;
        while (i >= 0 && result.get(i).equals("null")) {
            i--;
        }

        return result.subList(0, i + 1).toString();
    }

}

